### PR TITLE
improvement: handle client capabilities

### PR DESCRIPTION
**Pull request by camilledejoye**
The idea was that a client can tell us if it does handle snippet or not.
So we want to provide snippets only if the client can handle them.

I also took this opportunity for a bit a light refacto, nothing drastic and the tests covers it all.

### CODE REVIEW

**Comment by dantleech** [comment]
I thought we discussed making it part of the container parameters?

File: lib/LanguageServer/Util/ClientCapabilitiesProvider.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServer\Util;
+
+use InvalidArgumentException;
+use Phpactor\Extension\LanguageServer\Helper\ArrayHelper;
+use Psr\Container\ContainerInterface;
+
+final class ClientCapabilitiesProvider implements ContainerInterface
```
Comment (dantleech) [inline lib/LanguageServer/Util/ClientCapabilitiesProvider.php]: Why implement containerInterface?

**Comment by dantleech** [comment]
Nice idea -- but I'm not sure it's better than injecting the `bool` value - it's giving the class more information than it needs, and I think that injecting the client capabilities into the parameters _might_ be a way to go, but maybe not. I will have a think.

File: lib/LanguageServer/Util/ClientCapabilitiesProvider.php (hunk 1 above)
Comment (camilledejoye) [reply lib/LanguageServer/Util/ClientCapabilitiesProvider.php]: At first it looks like one, but I reverted my choice and forgot the interface.
I will remove it.

**Comment by camilledejoye** [comment]
> I thought we discussed making it part of the container parameters?

It's still part of the container, since I retrieve from the container to inject it into the provider.
I might have forgot something we discussed but I can't remember what exactly.

> Nice idea -- but I'm not sure it's better than injecting the bool value - it's giving the class more information than it needs, and I think that injecting the client capabilities into the parameters might be a way to go, but maybe not. I will have a think.

You mean injecting the flatten capabilities directly as parameter, so one parameter per capability ?
If so I'm not convinced, I've always seen parameter as a way to configure the container if that make sense.

I also thought it not the best solution to inject all capabilities at once, that why the in the previous version I injected only the relevant one.
But at then end I don't like to put the knowledge about the capabilities in the container, it's introducing a knowledge it should not care about. If we really want to inject only the relevant capabilities we should add a `HandlerFactory` that will deal with that knowledge.
But in our anonymous function are already some kind of "factories" but IMO they are here to lazy load more than to hold implementation knowledge.

**Comment by camilledejoye** [comment]
Was done in https://github.com/phpactor/language-server-phpactor-extensions/commit/11def7282072e6fc07f86ff42f6d29262be8d164

**Pull request closed without merging**